perfkitbenchmarker/providers/softlayer/softlayer_virtual_machine.py

```diff
@@ -0,0 +1,545 @@
+# Copyright 2020 PerfKitBenchmarker Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Class to represent an SoftLayer Virtual Machine object.
+
+All VM specifics are self-contained and the class provides methods to
+operate on the VM: boot, shutdown, etc.
+"""
+
+import json
+import logging
+import threading
+import string
+import random
+from time import sleep
+
+from perfkitbenchmarker import errors
+from perfkitbenchmarker import flags
+from perfkitbenchmarker import linux_virtual_machine
+from perfkitbenchmarker import virtual_machine
+from perfkitbenchmarker import vm_util
+from perfkitbenchmarker import windows_virtual_machine
+from perfkitbenchmarker.providers.softlayer import util
+from perfkitbenchmarker.providers.softlayer import softlayer_disk
+from perfkitbenchmarker import providers
+
+FLAGS = flags.FLAGS
+
+INSTANCE_EXISTS_STATUSES = frozenset(['ACTIVE'])
+INSTANCE_QUIESCING_TRANSACTION = frozenset(['CLOUD_RECLAIM_PREP'])
+INSTANCE_DELETED_STATUSES = frozenset(['DISCONNECTED'])
+INSTANCE_KNOWN_STATUSES = INSTANCE_EXISTS_STATUSES | INSTANCE_DELETED_STATUSES
+DRIVE_START_LETTER = 'c'
+
+
+class SoftLayerVirtualMachine(virtual_machine.BaseVirtualMachine):
+  """Object representing an SoftLayer Virtual Machine."""
+
+  CLOUD = providers.SOFTLAYER
+  DEFAULT_ROOT_DISK_TYPE = 'standard'
+
+
+  _lock = threading.Lock()  # ensure only one thread imports & exports sshkey
+  key_label = None
+  deleted_keyfile = False
+  imported_keyfile = False
+
+  def __init__(self, vm_spec):
+    """Initialize a SoftLayer virtual machine.
+
+    Args:
+      vm_spec: virtual_machine.BaseVirtualMachineSpec object of the vm.
+    """
+
+    super(SoftLayerVirtualMachine, self).__init__(vm_spec)
+    self.user_name = FLAGS.softlayer_user_name
```

Comment:
Flags should be attached to a config object in an _ApplyFlags method rather than accessed directly.